test/sdn_migration/sdn_migration_test.go

```diff
@@ -0,0 +1,639 @@
+					break
+				}
+			}
+		}
+
+		// Determine the appropriate action based on the upgrade state
+		switch upgradeState {
+		case "":
+			log.Printf("Upgrade has not started yet...")
+		case "Partial":
+			log.Printf("Upgrade is in progress. Conditions: %v", conditionMessage)
+		case "Completed":
+			log.Printf("Upgrade complete!")
+			return nil
+		case "Failed":
+			log.Printf("Upgrade failed! Conditions: %v", conditionMessage)
+			return &upgradeError{err: fmt.Errorf("upgrade failed")}
+		default:
+			log.Printf("Unknown upgrade state: %s", upgradeState)
+		}
+
+		// Wait before the next poll attempt
+		time.Sleep(upgradeDelay * time.Second)
+	}
+
+	return fmt.Errorf("upgrade is still in progress, failed to finish within max wait attempts")
+}
+
+// patchVersionConfig updates the version config to the desired version to initiate an upgrade
+func patchVersionConfig(ctx context.Context, client *openshiftclient.Client, channel string, image string, version string) error {
+	clusterVersionConfing := configv1.ClusterVersion{
+		TypeMeta:   v1.TypeMeta{},
+		ObjectMeta: v1.ObjectMeta{Name: "version"}}
+
+	mergePatch, err := json.Marshal(map[string]interface{}{
+		"spec": map[string]interface{}{
+			"channel": channel,
+			"desiredUpdate": map[string]interface{}{
+				"version": version,
+				"image":   image,
+				"force":   true,
+			},
+		},
+	})
+
+	if err != nil {
+		panic(err)
+	}
+
+	if err = client.Patch(
+		ctx,
+		&clusterVersionConfing,
+		k8s.Patch{PatchType: types.MergePatchType, Data: mergePatch},
+	); err != nil {
+		panic(err)
+	}
+	return nil
+
+}
+
+// patchNetworkConfig updates network type to OVN
+func patchNetworkConfig(ctx context.Context, client *openshiftclient.Client) error {
+
+	networkConfig := configv1.Network{ObjectMeta: v1.ObjectMeta{Name: "cluster"}}
+
+	mergePatch, err := json.Marshal(map[string]interface{}{
+		"metadata": map[string]interface{}{
+			"annotations": map[string]interface{}{
+				"network.openshift.io/network-type-migration": "",
+			},
+		},
+		"spec": map[string]interface{}{
+			"networkType": "OVNKubernetes",
+		},
+	})
+
+	if err != nil {
+		panic(err)
+	}
+
+	if err = client.Patch(
+		ctx,
+		&networkConfig,
+		k8s.Patch{PatchType: types.MergePatchType, Data: mergePatch},
+	); err != nil {
+		panic(err)
+	}
+	return nil
+}
+
+// addAnnotation adds the internal testing annotation to the network config
+func addIntenalTestingAnnotation(ctx context.Context, client *openshiftclient.Client) error {
+
+	networkConfig := configv1.Network{ObjectMeta: v1.ObjectMeta{Name: "cluster"}}
+
+	mergePatch, err := json.Marshal(map[string]interface{}{
+		"metadata": map[string]interface{}{
+			"annotations": map[string]interface{}{
+				"unsupported-red-hat-internal-testing": "true",
+			},
+		},
+	})
+
+	if err != nil {
+		panic(err)
+	}
+
+	if err = client.Patch(
+		ctx,
+		&networkConfig,
+		k8s.Patch{PatchType: types.MergePatchType, Data: mergePatch},
+	); err != nil {
+		panic(err)
```

Comment:
these panics should return err instead, there are a few uses of it in these helper functions